### PR TITLE
Extending DepResolver

### PR DESCRIPTION
**Pull request by kuritka**
related to (#170)

The primary purpose of this improvement is to read all k8gb input at one place, validate it and provide to GSLB in typed, well-formatted way. Depresolver is a single point of k8gb inputs now and fail-fast with proper error message if inputs are incorrect.

Inputs:

spec
 - DNSTtlSeconds
 - SplitBrainThresholdSeconds

config
 - ReconcileRequeueSeconds
 - ClusterGeoTag 
 - Route53Enabled
 - **ExtClustersGeoTags**
 - **EdgeDNSType**
 - **EdgeDNSServer**
 - **EdgeDNSZone**
 - **DNSZone**
-  **Infoblox.Host**
- **Infoblox.Version**
- **Infoblox.Port**
- **Infoblox.Username**
- **Infoblox.Password**

 
#### Validations
We implement the new validator, applying various rules to each input. 

### Tests
75% coverage checking depresolver works as expected. Extensive coverage allow us to make safe refactoring. 
Because I introduced new validation rules, I had to update controller_tests. 

### EdgeDNSType
Providing new computed property EdgeDNSType telling which configuration(s) loaded to depresolver.  This allows us to write conditions like 
```
if config.EdgeDNSType == DNSTypeRoute53 || config.EdgeDNSType == DNSTypeInfo
```
which overall simplify logic and helps in case we provide EdgeDNS implementations by Factory  

## depresolver as single point of environment variables names 
So we are not using names like "DNS_ZONE"  directly. Instead, we use depresolver.DnsZone constant.  I replaced variables in controller_tests while the DNS code will not access inputs directly. Will be part of following PR. 

## TODO:
Will follow in next PR (s). Thanks to validator, we can remove validations from GSLB logic and simplify code. Also, accessing depresolver instead of  direct access environment variables will enhance code readability. We can more concentrate on DNS logic rather than checking inputs. It will require changes in the code that I postpone to the next PR, because this PR is too large.

### CODE REVIEW

File: controllers/depresolver/depresolver.go [hunk 1]
```diff
@@ -14,17 +13,58 @@ import (
...
+	// EdgeDNSType is READONLY and is set automatically by configuration
+	EdgeDNSType EdgeDNSType
+	// EdgeDNSServer
+	EdgeDNSServer string
+	// EdgeDNSZone main zone which would contain gslb zone to delegate; e.g. example.com
+	EdgeDNSZone string
+	// DNSZone controlled by gslb; e.g. cloud.example.com
+	DNSZone string
+	// DNSTypeRoute53 switch
+	//TODO: hide for depresolver subscriber as depresolver retrieves EdgeDNSType. Maybe we can change configuration and set EdgeDNSType directly instead of DNSTypeRoute53 boolean
```
Comment (somaritane) [inline controllers/depresolver/depresolver.go:61]: Let's make comments consistent (i.e `// TODO` instead of `//TODO`)

Comment (kuritka) [reply controllers/depresolver/depresolver.go:61]: THX Tim, I fixed TODOs. I use TODO now as points that will need to resolve in following PR in order to not touch live code during super large PR.

**Review by somaritane — CHANGES_REQUESTED** [review]

File: controllers/depresolver/depresolver.go [hunk 2]
```diff
@@ -14,17 +13,58 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-//Config holds operator configuration
+// EdgeDNSType specifies to which edge DNS is k8gb connecting
+type EdgeDNSType int
+
+const (
+	// DNSTypeNoEdgeDNS is default DNSType. Is used during integration testing when no infoblox or route53 exists.
```
Comment (ytsarev) [inline controllers/depresolver/depresolver.go:20]:
```suggestion
	// DNSTypeNoEdgeDNS is default DNSType. Is used during integration testing when no edgeDNS provider exists.
```
Let's relax the explicit names here as we are about to add more providers in future

File: controllers/gslb_controller_test.go [hunk 3]
```diff
@@ -38,15 +40,29 @@ func TestGslbController(t *testing.T) {
...
 	if err != nil {
 		t.Fatalf("Can't setup env var: (%v)", err)
 	}
 
-	err = os.Setenv("FAKE_INFOBLOX", "true")
-	if err != nil {
-		t.Fatalf("Can't setup env var: (%v)", err)
+	predefinedConfig := depresolver.Config{
+		ReconcileRequeueSeconds: 30,
+		ClusterGeoTag:           "us-west1",
```
Comment (ytsarev) [inline controllers/gslb_controller_test.go:51]: Should be `us-west-1` format

File: controllers/depresolver/depresolver_validator.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+const (
+	//hostNameRegex allows cloud region formats; e.g. af-south-1
+	geoTagRegex = "^[a-zA-Z\\-\\d]*$"
```
Comment (ytsarev) [inline controllers/depresolver/depresolver_validator.go:11]: Are you sure we want to force it ? Also, how did you test the regex, didn't succeed to match region format for me, but I might be missing something

File: controllers/gslb_controller_test.go [hunk 5]
```diff
@@ -38,15 +40,29 @@ func TestGslbController(t *testing.T) {
...
+		ExtClustersGeoTags:      []string{"us-east1"},
```
Comment (ytsarev) [inline controllers/gslb_controller_test.go:52]: same as above

**Review by ytsarev — CHANGES_REQUESTED** [review]

File: controllers/depresolver/depresolver.go (hunk 1 above)
Comment (kuritka) [reply controllers/depresolver/depresolver.go]: @somaritane comments fixed

File: controllers/depresolver/depresolver_validator.go (hunk 4 above)
Comment (kuritka) [reply controllers/depresolver/depresolver_validator.go:11]: @ytsarev good point. Here is regex test:
[expected](https://github.com/AbsaOSS/k8gb/blob/170-include-envvars-into-depresolver/controllers/depresolver/depresolver_test.go#L228)
[corner case](https://github.com/AbsaOSS/k8gb/blob/170-include-envvars-into-depresolver/controllers/depresolver/depresolver_test.go#L217)

Comment (ytsarev) [reply controllers/depresolver/depresolver_validator.go:11]: perfect

**Review by ytsarev — APPROVED** [review]
Great stuff, thanks!

**Review by somaritane — APPROVED** [review]
